src/rulesets/Base/mapreduce.jl

```diff
@@ -322,3 +325,122 @@ end
     end
     return dx
 end
+
+#####
+##### `foldl`
+#####
+
+# `foldl` guarantees to execute `f` in order, left to right. So it makes sense even when
+# this `f` is stateful, in which case the gradient must be calculated in the reverse order. 
+# The implementation aims to be efficient for both tuples and arrays. 
+# Note that it does not return a gradient for `init`.
```

Comment:
I start to wonder if this package should add docstrings to functions, explaining what to expect, what pitfalls or alternatives to consider. Instead of just source comments.

Reply:
I often like attaching docstrings to internal functions, rather than comments.
For internal i think it doesn't matter as they are not part of the public API either way.
For `rrule` it is not quite the same, but still the big comments will not end up on website (which only uses ChainRulesCore) so it no big deal either way

Reply:
What I meant was adding docstrings to Base functions, so that hints about what AD is going to do / going to ignore / is good or bad at, are available as command line help.